src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,314 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import data.polynomial.algebra_map
+
+/-!  # Laurent polynomials
+
+Following a suggestion by Eric, Laurent polynomials are defined via `add_monoid_algebra R ℤ`.
+Thus, they are essentially `finsupp`s `ℤ →₀ R`.
+
+This means that they play well with polynomials, but there is a little roughness in establishing
+the API, since the `finsupp` implementation of `polynomial R` is well-shielded.
+
+## Notation
+The symbol `R[T;T⁻¹]` stands for `laurent_polynomial R`.
+
+## Implementation notes
+
+* `C : R →+* R[T;T⁻¹]` is the inclusion of constant polynomials, analogous to the one for `R[X]`;
+* `T : ℤ → R[T;T⁻¹]` is the sequence of powers of the variable `T`.
+
+Unlike the case of polynomials, I felt that the exponent notation was not too easy to use, as only
+natural exponents would be allowed.  Moreover, in the end, it seems likely that we should aim to
+performing computations on exponents in `ℤ` anyway and separating this via the symbol `T` seems
+convenient.
+
+I made some *heavy* use of `simp` lemmas, aiming to bring Laurent polynomials to the form
+`C a * T n`.  Any comments or suggestions for improvements is greatly appreciated!
+
+##  Future work
+Lots is missing!  I would certainly like to show that `R[T;T⁻¹]` is the localization of `R[X]`
+inverting `X`.  This should be mostly in place, given `exists_T_pow`.
+-/
+
+open_locale polynomial big_operators
+open polynomial add_monoid_algebra finsupp
+noncomputable theory
+
+variables {R N Z : Type*}
+
+/--  The semiring of Laurent polynomials with coefficients in the semiring `R`.
+We denote it by `R[T;T⁻¹]`.
+The ring homomorphism `C : R →+* R[T;T⁻¹]` includes `R` as the constant polynomials. -/
+abbreviation laurent_polynomial (R : Type*) [semiring R] := add_monoid_algebra R ℤ
+
+local notation R`[T;T⁻¹]`:9000 := laurent_polynomial R
+
+/--  The ring homomorphism, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent [semiring R] :
+  R[X] →+* R[T;T⁻¹] :=
+begin
+  refine ring_hom.comp _ (to_finsupp_iso R).to_ring_hom,
+  exact (map_domain_ring_hom R (nat.cast_add_monoid_hom ℤ)),
+end
+
+/--  The `R`-algebra map, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent_alg [comm_semiring R] :
+  R[X] →ₐ[R] R[T;T⁻¹] :=
+begin
+  refine alg_hom.comp _ (to_finsupp_iso_alg R).to_alg_hom,
+  exact (map_domain_alg_hom R R (nat.cast_add_monoid_hom ℤ)),
+end
+
+namespace laurent_polynomial
+
+section semiring
+variables [semiring R]
+
+lemma single_zero_one_eq_one : (single 0 1 : R[T;T⁻¹]) = (1 : R[T;T⁻¹]) := rfl
+
+/-!  ### The functions `C` and `T`. -/
+/--  The ring homomorphism `C`, including `R` into the ring of Laurent polynomials over `R` as
+the constant Laurent polynomials. -/
+def C : R →+* R[T;T⁻¹] :=
+{ to_fun    := single 0,
```

Comment:
Isn't there a `single_zero_ring_hom` you can use here?

Reply:
You are probably right: [lemma](https://leanprover-community.github.io/mathlib_docs/algebra/monoid_algebra/basic.html#add_monoid_algebra.single_zero_ring_hom)?

Reply:
Indeed, I used the lemma that you mentioned!